File: cses/03_dynamic_programming/1634_minimizing-coins/solution.py
import sys


def main():
    # Đọc toàn bộ dữ liệu một lần cho nhanh (vào/ra dạng buffer).
    data = sys.stdin.buffer.read().split()
    if not data:
        return
    n = int(data[0])
    x = int(data[1])
    coins = data[2:2 + n]

    # INF = giá trị "không tạo được". Vì mỗi đồng xu >= 1 nên đáp án thật luôn
    # <= x, do đó dùng x + 1 làm INF vừa an toàn vừa giữ số nhỏ.
    INF = x + 1

    # Loại đồng xu trùng nhau và bỏ đồng xu có giá trị > x (không dùng được).
    coin_vals = sorted({int(c) for c in coins if int(c) <= x})

    # dp[s] = số đồng xu ít nhất để tạo tổng s. Cơ sở: dp[0] = 0.
    dp = [INF] * (x + 1)
    dp[0] = 0

    # Unbounded knapsack: quét i tăng dần (vòng lặp xuôi) cho phép dùng lại một
    # đồng xu tùy ý. Chuyển: dp[i] = min(dp[i], dp[i - c] + 1).
    for c in coin_vals:
        d = dp  # Gán biến cục bộ để CPython truy cập nhanh hơn trong vòng lặp nóng.
        for i in range(c, x + 1):
            v = d[i - c] + 1
            if v < d[i]:
                d[i] = v

    # Nếu dp[x] vẫn > x (còn là INF) thì không tạo được, in -1.
    ans = dp[x]
    sys.stdout.write(str(ans if ans <= x else -1) + "\n")


if __name__ == "__main__":
    main()

File: cses/03_dynamic_programming/1634_minimizing-coins/solution.cpp
#include <algorithm>
#include <iostream>
#include <vector>

int main() {
    std::ios::sync_with_stdio(false);
    std::cin.tie(nullptr);

    int n;
    int target;
    std::cin >> n >> target;

    std::vector<int> coins(n);
    for (int& coin : coins) {
        std::cin >> coin;
    }

    // Đồng xu có giá trị > target không bao giờ dùng được; đồng xu trùng nhau
    // không đổi kết quả nên loại bớt cho gọn (sắp xếp, khử trùng, bỏ phần > target).
    std::sort(coins.begin(), coins.end());
    coins.erase(std::unique(coins.begin(), coins.end()), coins.end());
    coins.erase(std::upper_bound(coins.begin(), coins.end(), target),
                coins.end());

    // dp[s] = số đồng xu ít nhất để tạo tổng s. Khởi tạo INF = target + 1 để
    // đánh dấu "chưa tạo được"; đáp án thật luôn <= target nên không bị nhầm.
    const int unreachable = target + 1;
    std::vector<int> dp(target + 1, unreachable);
    dp[0] = 0;  // Cơ sở: tạo tổng 0 cần 0 đồng xu.

    // Unbounded knapsack: duyệt tổng tăng dần (vòng lặp xuôi) để mỗi đồng xu
    // được dùng lại tùy ý. Chuyển: dp[s] = min(dp[s], dp[s - coin] + 1).
    for (const int coin : coins) {
        for (int sum = coin; sum <= target; ++sum) {
            dp[sum] = std::min(dp[sum], dp[sum - coin] + 1);
        }
    }

    // Nếu dp[target] vẫn là INF thì không tạo được tổng target, in -1.

    std::cout << (dp[target] == unreachable ? -1 : dp[target]) << '\n';
    return 0;
}
